src/keys/slip10.rs

```diff
@@ -99,384 +188,196 @@ impl Seed {
+        self.key()
     }
 
-    pub fn child_key(&self, segment: &Segment) -> crate::Result<Self> {
-        #[cfg(feature = "ed25519")]
-        if self.0.curve == Curve::Ed25519 && !segment.is_hardened() {
-            return Err(crate::Error::InvalidArgumentError {
-                alg: "SLIP10",
-                expected: "hardened key index for Ed25519 master secret key",
-            });
-        }
-
-        Ok(Self(self.0.child_key(segment)))
-    }
-
-    pub fn derive(&self, chain: &Chain) -> crate::Result<Self> {
-        #[cfg(feature = "ed25519")]
-        if self.0.curve == Curve::Ed25519 && !chain.all_hardened() {
-            return Err(crate::Error::InvalidArgumentError {
-                alg: "SLIP10",
-                expected: "hardened key index for Ed25519 master secret key",
-            });
-        }
-
-        let mut key = self.0.clone();
-        for segment in &chain.0 {
-            key = key.child_key(segment);
-        }
-        Ok(Self(key))
+    pub fn into_extended_public_key(&self) -> Extended<K::PublicKey>
+    where
+        K::PublicKey: hazmat::IsPublicKey<SecretKey = K>,
+    {
+        Extended::from_extended_secret_key(self)
     }
+}
 
-    pub fn try_into_extended_public_key(&self) -> crate::Result<ExtendedPublicKey> {
-        if self.curve().is_non_hardened_supported() {
-            let mut k = KeyImpl {
-                curve: self.curve(),
-                ext: [0_u8; 65],
-            };
-            k.ext[..33].copy_from_slice(&self.0.calc_public_bytes());
-            k.ext[33..].copy_from_slice(self.chain_code());
-            Ok(ExtendedPublicKey(k))
-        } else {
-            Err(crate::Error::InvalidArgumentError {
-                alg: "SLIP10",
-                expected: "curve supporting non-hardened key derivation",
-            })
-        }
+impl<K: hazmat::IsSecretKey> From<&Seed> for Extended<K> {
+    fn from(seed: &Seed) -> Self {
+        Self::from_seed(seed)
     }
 }
 
-#[derive(Clone)]
-pub struct ExtendedPublicKey(KeyImpl);
-
-impl ExtendedPublicKey {
-    pub fn try_from_extended_bytes(curve: Curve, ext_bytes: &[u8; 65]) -> crate::Result<Self> {
-        if !curve.is_non_hardened_supported() {
-            Err(crate::Error::InvalidArgumentError {
-                alg: "SLIP10",
-                expected: "curve supporting non-hardened key derivation",
-            })
-        } else {
-            let key = KeyImpl { curve, ext: *ext_bytes };
-            if !key.is_public_key_valid() {
-                Err(crate::Error::InvalidArgumentError {
-                    alg: "SLIP10",
-                    expected: "valid extended public key",
-                })
-            } else {
-                Ok(Self(key))
-            }
-        }
+impl<K: hazmat::IsPublicKey> Extended<K> {
+    pub fn from_extended_secret_key(esk: &Extended<K::SecretKey>) -> Self {
+        let mut k = Self {
+            key: core::marker::PhantomData,
+            ext: [0_u8; 65],
+        };
```

Comment:
I see you doing this at different places, can it not be a default impl?

Reply:
It can't be Default - it wouldn't represent a valid inner key. I can factor it out into a private `new` method.

Reply:
Added `new` method.